kprobe inlined function

ttreyer opened the pull request:
Extend the `probe_inline` config to also apply to `kprobe`.

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

danobi reviewed (commented):
bit of early feedback: I don't think it's very common for vmlinux to come with dwarf. If available at all, it usually comes in a separate package.

ttreyer commented:
I reworked the initialisation of the DWARF parser to fail if the target doesn't have DebugInfo, so we don't query LLDB about a stripped binary. It improves the "Kernel DWARF" feature detection by not requiring a specific symbol to be present (an issue raised in #3445).
The calls to `lldb::SBDebugger::Initialize/Terminate` has also moved in its own class to benefit from RAII semantic. C++ object lifetime begins after the constructor returns, but if an exception is thrown in the constructor, the destructor won't be called. So if the `Dwarf` ctor threw an error, the `instance_counter` would not get decremented and `lldb::SBDebugger::Terminate()` would never get called.

jordalgo commented on .github/workflows/ci.yml:
```diff
@@ -82,6 +82,15 @@ jobs:
+      run: |
+        # From Ubuntu's Wiki: https://wiki.ubuntu.com/DebuggingProgramCrash#Non-built-in_debug_symbol_packages_.28.2A-dbgsym.29
+        sudo apt-get install ubuntu-dbgsym-keyring
+        echo -e "deb http://ddebs.ubuntu.com $(lsb_release -cs) main restricted universe multiverse\ndeb http://ddebs.ubuntu.com $(lsb_release -cs)-updates main restricted universe multiverse\ndeb http://ddebs.ubuntu.com $(lsb_release -cs)-proposed main restricted universe multiverse" | sudo tee -a /etc/apt/sources.list.d/ddebs.list
```
I love the 'restricted universe multiverse' part 😛

ttreyer replied:
haha
Ideally, we'd rely on vmtests to test on the Kernel.
Until then, we have to use this solution that can sometimes be flaky 🙁

danobi replied:
Looks like it takes about a minute and a half to install in CI. And given that CI stability and speed is fairly critical, not sure we should rush this in. Better to direct effort towards vmtest work instead if it's really that desired. 

Kernel dwarf is a tricky to use in open source world anyways - how do users even know to install the debug info anyways? Even if they want to, it's on the order of gigabytes and maximally tricky to install (as evidenced by this block). And that's for distros that even ship debug info.

I'm not objecting to the dwarf support itself - it is a contained change that can be useful if user really needs it. But the CI bit is the uncontained change - it affects the entire project on a daily basis. I'd much rather prefer to get a fast and reliable mechanism in place. 

Perhaps we can run a test like this internally? Should be much less flaky and still provide coverage.

jordalgo commented on tests/runtime/dwarf:
```diff
@@ -205,3 +205,10 @@ PROG struct Foo { int b; } uprobe:./testprogs/uprobe_test:uprobeFunction2 { prin
+
+NAME kprobe inlined function
+PROG config = { probe_inline = 1 } BEGIN { exit() }
+     kprobe:str_has_prefix { }
```
is this kernel function known to be commonly inlined?

ttreyer replied:
Yes, I looked in the Kernel source code for a function marked `always inlined` to test on.
Moreover, `str_has_prefix` seems like a stable target that won't disappear from the Kernel any time soon 🤞

jordalgo commented on tests/runtime/probe:
```diff
@@ -210,7 +210,7 @@ WILL_FAIL
 NAME kprobe_warn_missing_probes
 PROG kprobe:vfs_read,kprobe:nonsense { exit(); }
-EXPECT WARNING: could not attach probe kprobe:nonsense, skipping.
+EXPECT WARNING: could not attach probe kprobe:nonsense (kprobe:vfs_read,kprobe:nonsense), skipping.
```
why have the warning show both the probes if only one is skipped?

ttreyer replied:
Ah, I probably needed it for debugging the test. I can clean it up...

jordalgo commented on src/attached_probe.cpp:
```diff
@@ -524,78 +488,68 @@ bool AttachedProbe::resolve_offset_uprobe(bool safe_mode, bool has_multiple_aps)
-    LOG(V1) << "Could not resolve symbol " << symbol
+  auto path = find_vmlinux(&sym);
+  if (!path.has_value()) {
+    if (probe_.attach_point.empty())
```
```suggestion
    if (!is_symbol_kprobe)
```

jordalgo commented on src/attached_probe.cpp:
```diff
@@ -524,78 +488,68 @@ bool AttachedProbe::resolve_offset_uprobe(bool safe_mode, bool has_multiple_aps)
+  } else {
+    probe_.attach_point = std::move(sym.name);
+    if (__builtin_sub_overflow(probe_.address, sym.start, &probe_.func_offset))
+      throw FatalUserException("Offset before the function bounds ('" +
```
Maybe this is an opportunity to utilize the missing_probes config to either throw, log a warning, or just ignore and skip the probe.

jordalgo commented on src/attached_probe.cpp:
```diff
@@ -524,78 +488,68 @@ bool AttachedProbe::resolve_offset_uprobe(bool safe_mode, bool has_multiple_aps)
 
-  uint64_t sym_offset = resolve_offset(path, probe_.attach_point, probe_.loc);
+  if (probe_.func_offset >= sym.size)
+    throw FatalUserException("Offset outside the function bounds ('" +
```
same here

ttreyer replied:
I'm not sure it makes sense for both those cases to use the MissingProbes config.
1. The error above should be marked "unreachable". In `find_vmlinux(&sym)`, `libbcc` looks through each symbol and returns the one containing the given `probe_.address`. So we know that `probe_.address` is `>= sym.start`, otherwise we'd have failed at the `find_vmlinux` call. So this check is more about defensive programming.
2. The second one makes more sense to me, but we have to distinguish the case where the offset was give by the user or if it comes from multiple probes (like probing an inlined function). Moreover, it also need to be changed in `resolve_offset_uprobe`.
Let me know what you think and I'll update the PR accordingly 🙂

jordalgo replied:
For the error above #1, which you say is unreachable. Then it should probably just be a `LOG(BUG)` if we don't expect it to ever happen.

For #2. Yeah I agree. @ajor and I were sorta discussing similar around the alignment issue. We didn't really come to a consensus on how to handle the common case in terms of erroring or just warning and skipping attachment for the probe. We can push this to a discussion to a different PR though.

jordalgo commented on src/attached_probe.cpp:
```diff
@@ -524,78 +488,68 @@ bool AttachedProbe::resolve_offset_uprobe(bool safe_mode, bool has_multiple_aps)
+  uint64_t sym_offset = resolve_offset(path.value(),
+                                       probe_.attach_point,
+                                       probe_.loc);
+  check_alignment(path.value(),
```
This check was removed because the kernel will reject the attachment anyway.

jordalgo commented on src/utils.cpp:
```diff
@@ -109,6 +109,72 @@
+    // Format the loc.path with the current release.
+    char path[PATH_MAX] = {};
+    int bytes_written = std::snprintf(
+        path, sizeof(path), loc.path, uts.release);
```
Probably needs some kind of comment to ignore this. cc @danobi

jordalgo commented on src/utils.cpp:
```diff
@@ -109,6 +109,72 @@ const struct vmlinux_location vmlinux_locs[] = {
+        .use_symbol_type = BCC_SYM_ALL_TYPES ^ (1 << STT_NOTYPE),
+      };
+      if (bcc_elf_foreach_sym(path, callback, &options, sym) == -1) {
+        LOG(ERROR) << "Failed to iterate over symbols in " << path;
```
should this be `LOG(BUG)`?

ttreyer commented:
I don't understand why the runtime test `call.str_big_read` fails, but only for LLVM 15.
@danobi, do you have a clue?

jordalgo commented:
@ttreyer This may be broken or just incredibly flaky on trunk: https://github.com/bpftrace/bpftrace/pull/3503

ajor reviewed (changes requested):
I don't think we're actually attaching to the correct addresses here - `bpf_attach_kprobe` picks the address based on `funcname` and `offset_`. This PR sets `funcname` in the line above, so that's correct, but `offset_` is never updated to reflect `probe.address`.

As well as sorting this out, I think we need to add a way to validate that the addresses being attached to are correct so that we can check this in tests.

ttreyer commented on src/dwarf_parser.h:
```diff
@@ -43,6 +41,17 @@ class Dwarf {
+
+  public:
+    InstanceCounter();
+    virtual ~InstanceCounter();
```
Good point, it doesn't need to be, as this class is not inherited from.
I've remove the `virtual`, but marked the `InstanceCounter` class as `final` to prevent any inheritance.

ttreyer commented:
For the vulnerability with `snprintf` receiving its format string from the environment, we could require of the user to give us the path to the Kernel directly and let them call `uname -r` to have the correct path.

For example, on my machine it would look like:
```
# Before:
$ BFPTRACE_VMLINUX='/usr/lib/modules/%1$s/vmlinux-%1$s' ...

# After:
$ BPFTRACE_VMLINUX="/usr/lib/modules/$(uname -r)/vmlinux-$(uname -r)" ...
```

So if the `BPFTRACE_VMLINUX` environment variable is set, we can use it directly, without having to do any lookup to find the Kernel image.

ajor commented on tests/runtime/dwarf:
```diff
@@ -205,3 +205,14 @@ PROG struct Foo { int b; } uprobe:./testprogs/uprobe_test:uprobeFunction2 { prin
+# since they should be inline instances part of another function/symbol.
+NAME kprobe inlined function
+RUN {{BPFTRACE}} -ve \ 'config = { probe_inline = 1 } BEGIN { exit() } kprobe:str_has_prefix { }'
+EXPECT_NONE Attaching 2 probe...
```
```suggestion
EXPECT_NONE Attaching 2 probes...
```
Typo. Also do you think it's worth checking against "Attaching 1 probe..." as well?

ajor reviewed (approved):
Looks good!

ajor reviewed (changes requested):
@ttreyer After reading the discussion in #3504, I gave this program a test:
```
kprobe:__x64_sys_openat {}
```

With this PR, we now appear to be using DWARF even for non-inlined kprobes and we're attaching at `__x64_sys_openat+15` instead of `__x64_sys_openat+0`.

IMO non-inlined kprobes should always resolve addresses using the symbol table.
